docs/DesignDocs/shared_memory.md

```diff
@@ -0,0 +1,192 @@
+to TrustZone and any future platforms that employ a similar model. By forcing
+full serialization, the need for shared memory can be mostly eliminated. But
+some applications need to pass untrusted shared memory to enclaves. Examples
+include:
+
+- Ring buffers
+- Context-switchless calls
+- Efficient transfer of large memory objects
+
+Whatever the reasons, a mechanism is needed for sharing untrusted memory with
+the enclave and that mechanism must be interoperable across secure hardware
+platforms. Further, the mechanism must be explicit, so operations are needed
+for managing untrusted memory.
+
+User Experience
+---------------
+
+The host may share untrusted memory with the enclave through the use of the
+**user_check** EDL attribute. For example:
+
+```c++
+enclave
+{
+    trusted
+    {
+        void share_buffer_ecall(
+            [user_check] void* untrusted_memory,
+            size_t untrusted_memory_size);
+    };
+};
+```
+
+We assume in this example that the enclave needs to read or write the
+**untrusted_memory** buffer. Prior to this proposal, SGX applications
+could pass any pointer to untrusted memory. For example:
+
+```c++
+    static unsigned char buf[4096];
+
+    share_buffer_ecall(buf, sizeof(buf));
+```
+
+But this application is not portable to TrustZone since **buf** is not visible
+within the enclave. To make this work on TrustZone, the buffer should be
+explicitly allocated from a shared memory pool as follows.
+
+```c++
+    const size_t n = 4096;
+    unsigned char* buf;
+
+    if (oe_allocate_shared_memory(enclave, n, &buf) != OE_RESULT_OK)
+    {
+        /* Error */
+    }
+
+    memset(buf, 0, n);
+
+    share_buffer_ecall(buf, n);
+```
+
+This program fragment will work on both SGX and TrustZone. The application
+will eventually free this buffer by passing it to **oe_free_shared_memory()**.
+
+The enclave's implementation of **share_buffer_ecall()** should verify that
+the **untrusted_memory** is fully contained within shared memory as shown
+below.
+
+```c++
+void share_buffer_ecall(void* untrusted_memory, size_t untrusted_memory_size)
+{
+    if (!oe_is_shared_memory(untrusted_memory, untrusted_memory_size)
+    {
+        /* Error! */
+    }
+    ...
+}
+```
+
+Specification
+-------------
+
+This design introduces the following functions on the host side:
+
+```c++
+    oe_result_t oe_allocate_shared_memory(
+        oe_enclave_t* enclave,
+        size_t size,
+        void** ptr);
+
+    oe_result_t oe_free_shared_memory(
+        oe_enclave_t* enclave,
+        void* ptr);
+
+    bool oe_is_shared_memory(
+        oe_enclave_t* enclave,
+        const void* ptr,
+        size_t size);
+
+    typedef enum _oe_memory_type
+    {
+        /* !oe_is_within_enclave() && !oe_is_shared_memory() */
+        OE_MEMORY_TYPE_UNKNOWN,
+
+        /* oe_is_within_enclave() */
+        OE_MEMORY_TYPE_TRUSTED,
+
+        /* oe_is_shared_memory() */
+        OE_MEMORY_TYPE_SHARED,
+    }
```

Comment:
Worth defining is a portion of memory will always be shared or not.
From what I understand, in SGX 2, an enclave can grow by adding TCS pages after creation. Does that mean that a memory that was once shared can become enclave-memory if it is freed and then rellocated to the enclave?